FI-2732: Add SMART App Launch capability (#9)
inferno-framework/davinci-dtr-test-kit: lib/davinci_dtr_test_kit/docs/dtr_smart_app_suite_description_v201.md

### Quick Start

This test suite can be run in two modes, each described below:
1. [EHR launch mode](#ehr-launch)
2. [Standalone launch mode](#standalone-launch)

At this time, Inferno's simulation of the payer server that provides the questionnaires

> uses the same base server url and access token and apps will need to be configured to

Author: ```suggestion
uses the same base server url and access token, and apps will need to be configured to
```

> connect to it as well. See [here](#combined-payer-and-ehr-fhir-servers) for details.

Author: This anchor link doesn't work, because the headings in this markdown don't get IDs

```suggestion
connect to it as well. See the "Combined payer and EHR FHIR servers" section below for details.
```

The DTR specification allows apps and their partners significant leeway in how they decide

> what questionnaire to request. Inferno cannot know ahead of time what data needs to be

Author: They're not requesting the questionnaire per se though, right? Would it be more accurate to say that the spec "allows them leeway in how they determine which payer and endpoint from which to request the questionnaire package."

Reviewer: I took a shot at clarifying this - see what you think. They are requesting a questionnaire, but you're right that it is more than that.

available for the app under test to successfully request, pre-populate, and render 

> a questionnaire. See the [`fhirContext` and available instances](#fhircontext-and-available-instances)

Author: Same issue

```suggestion
a questionnaire. See the "`fhirContext` and available instances"
```

section below for details on how to enable Inferno to meet the needs of your application.

#### EHR Launch

In this mode Inferno will launch the app under test using the SMART App Launch

> [EHR launch](https://hl7.org/fhir/smart-app-launch/app-launch.html#launch-app-ehr-launch)

Author: Should be STU2.1 link

```suggestion
[EHR launch](https://hl7.org/fhir/smart-app-launch/STU2.1/app-launch.html#launch-app-ehr-launch)
```

flow. 

The tester must provide
1. a `client_id`: can be any string and will uniquely identify the testing session.

> 2. a `launch_uri`: will be used by Inferno to launch the app under test.

Author: ```suggestion
2. a `launch_uri`: will be used by Inferno to generate a link to launch the app under test.
```

> All the details needed to access clinical data from Inferno's simulated are provided

Author: ```suggestion
All the details needed to access clinical data from Inferno's simulated EHR are provided
```

as a part of the SMART flow, including 

> - the FHIR base server url to request data from

Author: ```suggestion
- the FHIR base server URL to request data from
```

- a bearer token to provide on all requests

#### Standalone Launch

In this mode the app under test will launch and on its own and reach out to Inferno to
begin the workflow as described in the 

> [standalone launch section](https://hl7.org/fhir/smart-app-launch/app-launch.html#launch-app-standalone-launch).

Author: Should be STU2.1 link

```suggestion
[standalone launch section](https://hl7.org/fhir/smart-app-launch/STU2.1/app-launch.html#launch-app-standalone-launch).
```

## Configuration Details

### `fhirContext` and available instances

Once they have launched, DTR SMART Apps obtain details that drive their retrieval of questionnaires
and relevant clinical data from the payer and the EHR from [context that is passed with
the access token](https://hl7.org/fhir/smart-app-launch/STU2.1/scopes-and-launch-context.html)
provided by the EHR. Inferno cannot know ahead of time what information to provide and
what instances to make available to direct the app under test to request and render a
particular questionnaire.

Therefore, use of this test suite requests that the tester provide this information so that the
app can demonstrate its capabilities based on whatever business logic is present. These tests
currently support 2 context parameters that contain references to instance in the EHR and provides
testers with a way to provide those instances to Inferno so it can serve them to the app. These are
controlled by the following inputs present on each group associated with a questionnaire:

> - `SMART App Launch Patient ID`: provide an id for the subject Patient FHIR instance.
> - `SMART App Launch fhirContext`: provide a json object containing FHIR references to instances

Author: ```suggestion
- `SMART App Launch Patient ID`: provide an `id` for the subject Patient FHIR instance.
- `SMART App Launch fhirContext`: provide a JSON object containing FHIR references to instances
```

>   relevant to the DTR workflow, e.g. 
>   `[{reference: 'Coverage/cov015'}, {reference: 'DeviceRequest/devreqe0470'}]`. This will be included
>   under the `fhirContext` key of the token response.

Author: This is hard to read

```suggestion
  relevant to the DTR workflow, e.g. 

  ```
  [{reference: 'Coverage/cov015'}, {reference: 'DeviceRequest/devreqe0470'}]
  ```

  This will be included under the `fhirContext` key of the token response.
```

Reviewer: committed this change

> - `SMART App Launch Patient ID`: provide an id for the subject Patient FHIR instance.
> - `SMART App Launch fhirContext`: provide a json object containing FHIR references to instances
>   relevant to the DTR workflow, e.g. 
>   `[{reference: 'Coverage/cov015'}, {reference: 'DeviceRequest/devreqe0470'}]`. This will be included
>   under the `fhirContext` key of the token response.
> - `EHR-available resources`: provide a Bundle containing FHIR instances referenced in and from the

Author: Bold instead of fixed width

```suggestion
- **SMART App Launch Patient ID**: provide an id for the subject Patient FHIR instance.
- **SMART App Launch `fhirContext`**: provide a json object containing FHIR references to instances
  relevant to the DTR workflow, e.g. 
  `[{reference: 'Coverage/cov015'}, {reference: 'DeviceRequest/devreqe0470'}]`. This will be included
  under the `fhirContext` key of the token response.
- **EHR-available resources**: provide a Bundle containing FHIR instances referenced in and from the
```

Reviewer: committed this change

>   previous two inputs. Each instance must include a `id` element that Inferno will use in conjunction

Author: ```suggestion
  previous two inputs. Each instance must include an `id` element that Inferno will use in conjunction
```

where you can find `id` and `url` values and any other details needed to determine what inputs 
will allow the app under test to work with that questionnaire. Note additionally that Inferno will always
return that questionnaire in response to `$questionnaire-package` requests made during that test.

These inputs can be cumbersome to create and if you have suggestions about how to improve this process
while keeping the flexibility of Inferno to run with any app, submit a ticket 
[here](https://github.com/inferno-framework/davinci-pas-test-kit/issues).

## Limitations

The DTR IG is a complex specification and these tests currently validate conformance to only
a subset of IG requirements. Future versions of the test suite will test further
features. A few specific features of interest are listed below.

### Launching and security

This test kit contains basic SMART App Launch cabilities that may not be complete. In particular,
refresh tokens are not currently supported and scopes are not precise. To provide feedback and 
input on the design of this feature and help us priortize improvements, submit a ticket 
[here](https://github.com/inferno-framework/davinci-pas-test-kit/issues).

> ### Combined payer and ehr fhir servers

Author: ```suggestion
### Combined payer and EHR FHIR servers
```

> At this time, the test suite simulates a single FHIR server that uses the same access token
> for both the payer server and the ehr server. Apps under test must use theFHIR base server url
> and access token identified during the SMART app launch sequence when making requests

Author: ```suggestion
At this time, the test suite simulates a single FHIR server that uses the same access token
for both the payer server and the EHR server. Apps under test must use the FHIR server base url
and access token identified during the SMART App Launch sequence when making requests
```
